### PR TITLE
Improve module, type and function definitions

### LINKED ISSUE #26: Anonymous functions are not highlighted correctly
The `function` keyword for the following anonymous function is not highlighted correctly:

image

### LINKED ISSUE #36: short function definition syntax
The short function definition
```julia
f(x, y) = x + y
```
is parsed as a call expression with arguments, not as a function definition with parameters.
```
source_file [0, 0] - [1, 0]
  assignment_expression [0, 0] - [0, 15]
    call_expression [0, 0] - [0, 7]
      identifier [0, 0] - [0, 1]
      argument_list [0, 1] - [0, 7]
        identifier [0, 2] - [0, 3]
        identifier [0, 5] - [0, 6]
    operator [0, 8] - [0, 9]
    binary_expression [0, 10] - [0, 15]
      identifier [0, 10] - [0, 11]
      operator [0, 12] - [0, 13]
      identifier [0, 14] - [0, 15]
```

### LINKED ISSUE #40: Update the treesitter version to 0.20.0
I did commend on this in the 'project status' issue
Some new editors now work with tree-sitter 0.20.0 release and using this version of tree-sitter-julia there isn't feasible in my experience. I did try bumping the version in `Cargo.toml` to try with a an editor but that broke some stuff.
How much work would it take to port tree-sitter-julia to tree-sitter 0.20.0?

### LINKED ISSUE #41: Fail to parse: single line struct
While playing around with tree sitter I came across some code that fails to parse currently. While I don't have the skills to fix them yet, I hope isolating them and opening individual issues will help someone in troubleshooting these. 

_Source_
```
struct ThisThing; end
```

_Parse result_
```
(source_file [0, 0] - [0, 21]
  (struct_definition [0, 0] - [0, 21]
    name: (identifier [0, 7] - [0, 16])
    (ERROR [0, 16] - [0, 17])))
```

When the struct definition is split into two lines, the parsing succeeds. 

_Source_
```
struct ThisThing
end
```

_Parse result_
```
(source_file [0, 0] - [0, 21]
  (struct_definition [0, 0] - [0, 21]
    name: (identifier [0, 7] - [0, 16])
    (ERROR [0, 16] - [0, 17])))
```

### LINKED ISSUE #42: Fail to parse: Module in function definition
_Source_
```julia
function Base.getproperty(x, y)
	return 0
end
```

_Parse Result_
```
(source_file [0, 0] - [2, 3]
  (function_definition [0, 0] - [2, 3]
    (ERROR [0, 9] - [0, 14]
      (identifier [0, 9] - [0, 13]))
    name: (identifier [0, 14] - [0, 25])
    parameters: (parameter_list [0, 25] - [0, 31]
      (identifier [0, 26] - [0, 27])
      (identifier [0, 29] - [0, 30]))
    (return_statement [1, 1] - [1, 9]
      (integer_literal [1, 8] - [1, 9]))))
```

### LINKED ISSUE #57: Parse error parsing function definition without a named argument
Source:
```julia
function foo(::Int)
end
```
Playground:
```
function_definition [0, 0] - [1, 3]
  name: identifier [0, 9] - [0, 12]
  parameters: parameter_list [0, 12] - [0, 19]
    ERROR [0, 13] - [0, 15]
    identifier [0, 15] - [0, 18]
```
Note that it works correctly for short form definitions (`foo(::Int) = nothing`).

### LINKED ISSUE #60: Error parsing parametric method without {} around parameter
If you only have one parameter it is optional to use curly braces. Source:
```julia
function foo(x::T) where T
end
```
Playground:
```
function_definition [0, 0] - [1, 3]
  name: identifier [0, 9] - [0, 12]
  parameters: parameter_list [0, 12] - [0, 18]
    typed_parameter [0, 13] - [0, 17]
      identifier [0, 13] - [0, 14]
      identifier [0, 16] - [0, 17]
  identifier [0, 19] - [0, 24]
  ERROR [0, 25] - [0, 26]
    identifier [0, 25] - [0, 26]
```

### LINKED ISSUE #61: Error parsing prefixed string literal in function definition
Source:
```julia
function foo(x::MIME"text/plain")
end
```
Playground:
```
function_definition [0, 0] - [1, 3]
  name: identifier [0, 9] - [0, 12]
  parameters: parameter_list [0, 12] - [0, 33]
    identifier [0, 13] - [0, 14]
    ERROR [0, 14] - [0, 32]
      prefixed_string_literal [0, 16] - [0, 32]
        prefix: identifier [0, 16] - [0, 20]
```

### LINKED ISSUE #63: Error parsing splat variable with type
Source:

```
function foo(bar::T...)
  bar
end
```

Playground:

```
ERROR [0, 0] - [2, 3]
  identifier [0, 9] - [0, 12]
  identifier [0, 13] - [0, 16]
  field_expression [0, 18] - [1, 4]
    identifier [0, 18] - [0, 19]
    ERROR [0, 20] - [0, 23]
    identifier [1, 1] - [1, 4]
  identifier [2, 0] - [2, 3]
```

### PR DESCRIPTION
**Pull request by savq**
**Types and modules**
- Add bare module definitions.
- Fix single line structs.
- Split tests for structs and other types.

**Functions**
- Add operators as names in function definitions.
- Add zero method function definitions.
- Add short form function definitions.
- Add function signature rule (shared by block and short function definitions).
- Add `where` clauses.
- Refactor anonymous block functions (`function() ... end`). Consider them function definitions, not function expressions.
- Split tests for function definitions into 4 sections.

*Function parameters and scoped identifiers*
- Add anonymous typed parameters.
- Add parameters to `do_clause`.
- Rename `spread_expression` to `slurp_expression` (yes, that's the real name).
- Allow `_primary_expression` and `prefixed_string_literal` as types in `typed_parameter` and `return_type`.
- Allow scoped identifiers in:
  - function definitions
  - macro identifiers
  - type parameters

**Misc**
- Split import rules into `selected_import` and `_import_list`. These cannot be used in the same import statement.
- Remove 29 files from known-failures.

**Caveats:**
- `where` is actually a binary operator. It can be used in places beyond function definitions. A new rule could be added in the future to handle more use cases.
- I wanted to add a `_parameter` rule as suggested [here](https://github.com/tree-sitter/tree-sitter-julia/pull/47#issuecomment-1233295813), but this caused more conflicts than necessary.
- I made the terminator after `do_clause` parameters mandatory.
- The conflicts that were added are mostly due to the fact that `assignment_expression` still allows `call_expression` in the LHS.
- The reason for calling `...` a "slurp" comes from here: [What does the `...` operator do?](https://docs.julialang.org/en/v1/manual/faq/#What-does-the-...-operator-do?)

Once assignment patterns are fixed, these conflicts could be removed (and `_parameter` could finally be added), but I think that belongs in another PR.

---
Closes #26
Closes #40 
Closes #41 
Closes #42
Closes #36, but the implementation is different from #43
Closes #57
Closes #60 
Closes #61
Closes #63

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -105,13 +105,21 @@ grammar({
   ],
 
   conflicts: $ => [
-    // Arrow functions vs tuples
+    [$._primary_expression, $._function_signature],
     [$._primary_expression, $.parameter_list],
     [$._primary_expression, $.spread_parameter],
```
Comment (ChrHorn) [inline grammar.js:110]: dwd

**Comment by ChrHorn** [comment]
Very nice, I will close https://github.com/tree-sitter/tree-sitter-julia/pull/43 in favor of getting this one through.

Could you also add support for empty functions, for example `function f end`. These are often used to add docstrings.

File: test/corpus/definitions.txt [hunk 2]
```diff
@@ -117,37 +123,128 @@ end
...
+==============================
+
+function f(x::T)
+end
+
+function g(n::N) where {n <: Integer}
+    n
+end
+
+g(n::N) where {n <: Integer} = n
```
Comment (pabloferz) [inline test/corpus/definitions.txt:200]:
```suggestion
function f(x::T) where {T}
end

function g(n::N) where {N <: Integer}
    n
end

g(n::N) where {N <: Integer} = n
```

**Comment by ExpandingMan** [comment]
This is awesome, thanks!

Just wanted to report a minor issue I found in which sometimes type assertions can be parsed as symbols, see this example
image

**Comment by ExpandingMan** [comment]
Another minor issue, the `∈` character in `for` loops is not being recognized (should be `@keyword.operator` to match `in`).

image

File: grammar.js [hunk 3]
```diff
@@ -183,20 +191,53 @@ grammar({
...
+      $._expression,
+    )),
+
+    _function_signature: $ => seq(
+      field('name', choice(
+        $.identifier,
+        $.operator,
+        $.scoped_identifier,
+      )),
+      field('type_parameters', optional($.type_parameter_list)),
```
[the remark this answers is not in the document]
Comment (savq) [reply grammar.js:232]: I'm going to leave like that for now and check it again when refactoring primary expressions.

File: test/corpus/definitions.txt (hunk 2 above)
Comment (savq) [reply test/corpus/definitions.txt]: Whoops. Fixed.

**Comment by savq** [comment]
> Just wanted to report a minor issue I found in which sometimes type assertions can be parsed as symbols, see this example

@ExpandingMan The PR fixes this 👍

> Another minor issue, the ∈ character in for loops is not being recognized (should be @keyword.operator to match in).

This can be fixed in either the grammar or the queries, and it's a bit unrelated to the rest of the PR. I'll add a fix for this in nvim-treesitter first.

**Comment by ChrHorn** [comment]
Regarding the renaming of `spread_expression` to `slurp_expression`:

I would prefer to call it `splat_expression`.
The docstring of `...` calls it the "splat operator" and the [docs](https://docs.julialang.org/en/v1/manual/functions/) mention "splat" four times, but "slurp" only once.

**Comment by savq** [comment]
Ok. All the definitions and their correspoding tests should be ready.

File: grammar.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (savq) [reply grammar.js:232]: For future reference:
- [parametric composite types](https://docs.julialang.org/en/v1/manual/types/#man-parametric-composite-types)
- [parametric constructors](https://docs.julialang.org/en/v1/manual/constructors/#Parametric-Constructors)

**Comment by savq** [comment]
Added fix to close #60

**Comment by tecosaur** [comment]
Regarding `slurp_expression`, would `splat_expression` be more appropriate? I see `...` described as the "splat operator" more than anything else. You'll also see it called the "splat operator" (but not the "slurp operator") in Python, Ruby, and other places.

**Comment by savq** [comment]
I wrote the wrong thing in the PR comment (I updated it). For future reference:

- `spread_parameter` was renamed to `slurp_parameter` (here)
- `spread_expression` was renamed to `splat_expression` (in #83)

That's the "correct" naming convention in Julia. See [What does the `...` operator do?](https://docs.julialang.org/en/v1/manual/faq/#What-does-the-...-operator-do?)

**Pull request merged**